### PR TITLE
Return the current position & leaf from the witness operation.

### PR DESCRIPTION
**Pull request by nuttycom**

### CODE REVIEW

**Review by therealyingtong — APPROVED** [review]
utACK

File: src/bridgetree.rs [hunk 1]
```diff
@@ -879,9 +879,10 @@ impl<H: Hashable, const DEPTH: u8> Tree<H> for BridgeTree<H, DEPTH> {
     }
 
     /// Marks the current tree state leaf as a value that we're interested in
-    /// witnessing. Returns true if successful and false if the tree is empty.
-    fn witness(&mut self) -> bool {
-        if let Some(key @ (position, _)) = self.current_leaf() {
+    /// witnessing. Returns the current position and leaf value if the treee
```
Comment (str4d) [inline src/bridgetree.rs:882]:
```suggestion
    /// witnessing. Returns the current position and leaf value if the tree
```

File: src/lib.rs [hunk 2]
```diff
@@ -245,11 +245,11 @@ pub trait Tree<H>: Frontier<H> {
...
-    /// Marks the current leaf as one for which we're interested in producing
-    /// an authentication path. Returns an optional value containing the current
-    /// position if successful or if the current value was already marked,
-    /// or None if the tree is empty.
-    fn witness(&mut self) -> bool;
+    /// Marks the current leaf as one for which we're interested in producing 
+    /// an authentication path. Returns an optional value containing the 
+    /// current position and leaf value if successful or if the current 
+    /// value was already marked, or None if the tree is empty.
+    fn witness(&mut self) -> Option<(Position, H)>;
```
Comment (str4d) [inline src/lib.rs:252]: I was going to ask for this to be documented in the changelog, but apparently we don't have one for this crate yet, so I'm not going to block on this. Open an issue for going through the API changes across crate releases and documenting them.

File: src/lib.rs [hunk 3]
```diff
@@ -245,11 +245,11 @@ pub trait Tree<H>: Frontier<H> {
     /// the specified leaf value from the specified position in the tree.
     fn is_witnessed(&self, position: Position, value: &H) -> bool;
 
-    /// Marks the current leaf as one for which we're interested in producing
-    /// an authentication path. Returns an optional value containing the current
-    /// position if successful or if the current value was already marked,
-    /// or None if the tree is empty.
-    fn witness(&mut self) -> bool;
+    /// Marks the current leaf as one for which we're interested in producing
```
Comment (str4d) [inline src/lib.rs:248]:
```suggestion
    /// Marks the current leaf as one for which we're interested in producing
```
I also don't understand why the documentation change for this function is present: it seems like the documentation was incorrect before but correct now?

File: src/lib.rs [hunk 4]
```diff
@@ -782,8 +782,9 @@ pub(crate) mod tests {
         }
 
         /// Marks the current tree state leaf as a value that we're interested in
-        /// witnessing. Returns true if successful and false if the tree is empty.
-        fn witness(&mut self) -> bool {
+        /// witnessing. Returns the current position and leaf value if the treee
```
Comment (str4d) [inline src/lib.rs:785]:
```suggestion
        /// witnessing. Returns the current position and leaf value if the tree
```

File: src/sample.rs [hunk 5]
```diff
@@ -65,17 +65,15 @@ impl<H: Hashable + PartialEq + Clone> TreeState<H> {
     }
 
     /// Marks the current tree state leaf as a value that we're interested in
-    /// witnessing. Returns true if successful and false if the tree is empty.
-    fn witness(&mut self) -> bool {
-        if let Some((pos, value)) = self.current_leaf() {
+    /// witnessing. Returns the current position and leaf value if the treee
```
Comment (str4d) [inline src/sample.rs:68]:
```suggestion
    /// witnessing. Returns the current position and leaf value if the tree
```

File: src/sample.rs [hunk 6]
```diff
@@ -207,8 +205,9 @@ impl<H: Hashable + PartialEq + Clone> Tree<H> for CompleteTree<H> {
     }
 
     /// Marks the current tree state leaf as a value that we're interested in
-    /// witnessing. Returns true if successful and false if the tree is empty.
-    fn witness(&mut self) -> bool {
+    /// witnessing. Returns the current position and leaf value if the treee
```
Comment (str4d) [inline src/sample.rs:208]:
```suggestion
    /// witnessing. Returns the current position and leaf value if the tree
```

**Review by str4d — CHANGES_REQUESTED** [review]

**Review by str4d — APPROVED** [review]
utACK after above documentation fixes.

**Pull request merged**